sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3658,6 +3658,81 @@ pi_result piQueueGetInfo(pi_queue Queue, pi_queue_info ParamName,
   case PI_QUEUE_INFO_DEVICE_DEFAULT:
     die("PI_QUEUE_INFO_DEVICE_DEFAULT in piQueueGetInfo not implemented\n");
     break;
+  case PI_EXT_ONEAPI_QUEUE_INFO_STATUS: {
+    // We can exit early if we have in-order queue.
+    if (Queue->isInOrderQueue()) {
+      if (!Queue->LastCommandEvent)
+        return ReturnValue(pi_bool{true});
+
+      // We can check status of the event only if it isn't discarded otherwise
+      // it may be reset (because we are free to reuse such events) and
+      // zeEventQueryStatus will hang.
+      if (!Queue->LastCommandEvent->IsDiscarded) {
+        ze_result_t ZeResult = ZE_CALL_NOCHECK(
+            zeEventQueryStatus, (Queue->LastCommandEvent->ZeEvent));
+        if (ZeResult == ZE_RESULT_NOT_READY) {
+          return ReturnValue(pi_bool{false});
+        } else if (ZeResult != ZE_RESULT_SUCCESS) {
+          return mapError(ZeResult);
+        }
+        return ReturnValue(pi_bool{true});
+      }
+      // For immediate command lists we have to check status of the event
+      // because immediate command lists are not associated with level zero
+      // queue. Conservatively return false in this case because last event is
+      // discarded and we can't check its status.
+      if (Queue->Device->useImmediateCommandLists())
+        return ReturnValue(pi_bool{false});
```

Comment:
I am afraid that it is not OK to return "false" here because there would be no forward progress.
Can we instead iterate over all events in the imm command-list as you do below?

Reply:
Sorry but I don't completely understand why it is not ok to return false and why there would be no forward progress, can you please clarify that?
Unfortunately we can't check status of the discarded events because they may be reset.

Reply:
I mean every subsequent query of such queue (immediate submission with discarded events) would have to return false.

Reply:
If queue or event is synchronized then true will be returned.

Reply:
> If queue or event is synchronized then "true" will be returned.

I see, this is because `LastCommandEvent` will be null. OK